File: hooks/scripts/internal/notify/discord.go
// discord.go: Discord webhook notification provider using rich embeds.
package notify

import (
	"path/filepath"
	"time"
)

var discordColors = map[string]int{
	"Stop":          5763719,  // Green
	"SubagentStop":  3447003,  // Blue
	"AskUserPrompt": 15844367, // Yellow
	"default":       10070709, // Gray
}

// Discord is the Discord webhook provider.
var Discord = Provider{
	Name: "discord",
	IsEnabled: func(env map[string]string) bool {
		return env["DISCORD_WEBHOOK_URL"] != ""
	},
	Send: func(input map[string]interface{}, env map[string]string) (*SendResult, error) {
		webhookURL := env["DISCORD_WEBHOOK_URL"]
		if webhookURL == "" {
			return &SendResult{Error: "DISCORD_WEBHOOK_URL not configured"}, nil
		}
		embed := buildDiscordEmbed(input)
		body := map[string]interface{}{
			"embeds": []interface{}{embed},
		}
		return Send("discord", webhookURL, body, nil), nil
	},
}

// discordField is a Discord embed field.
type discordField struct {
	Name   string `json:"name"`
	Value  string `json:"value"`
	Inline bool   `json:"inline"`
}

// discordEmbed is a Discord message embed.
type discordEmbed struct {
	Title       string         `json:"title"`
	Description string         `json:"description"`
	Color       int            `json:"color"`
	Timestamp   string         `json:"timestamp"`
	Footer      map[string]string `json:"footer"`
	Fields      []discordField `json:"fields"`
}

// buildDiscordEmbed constructs the embed payload for a hook input.
func buildDiscordEmbed(input map[string]interface{}) discordEmbed {
	hookType := strField(input, "hook_event_name", "unknown")
	cwd := strField(input, "cwd", "")
	sessionID := strField(input, "session_id", "")

	projectName := "Unknown"
	if cwd != "" {
		if b := filepath.Base(cwd); b != "" {
			projectName = b
		}
	}

	color, ok := discordColors[hookType]
	if !ok {
		color = discordColors["default"]
	}

	sessionDisplay := "`N/A`"
	if len(sessionID) >= 8 {
		sessionDisplay = "`" + sessionID[:8] + "...`"
	} else if sessionID != "" {
		sessionDisplay = "`" + sessionID + "`"
	}

	ts := time.Now().Format("15:04:05")
	isoTS := time.Now().UTC().Format(time.RFC3339)
	cwdDisplay := cwd
	if cwdDisplay == "" {
		cwdDisplay = "Unknown"
	}
	locationField := discordField{Name: "📍 Location", Value: "`" + cwdDisplay + "`", Inline: false}

	var title, description string
	var fields []discordField

	switch hookType {
	case "Stop":
		title = "Claude Code Session Complete"
		description = "Session completed successfully"
		fields = []discordField{
			{Name: "⏰ Time", Value: ts, Inline: true},
			{Name: "🆔 Session", Value: sessionDisplay, Inline: true},
			locationField,
		}
	case "SubagentStop":
		agentType := strField(input, "agent_type", "unknown")
		title = "Claude Code Subagent Complete"
		description = "Specialized agent completed its task"
		fields = []discordField{
			{Name: "⏰ Time", Value: ts, Inline: true},
			{Name: "🔧 Agent Type", Value: agentType, Inline: true},
			{Name: "🆔 Session", Value: sessionDisplay, Inline: true},
			locationField,
		}
	case "AskUserPrompt":
		title = "Claude Code Needs Input"
		description = "Claude is waiting for user input"
		fields = []discordField{
			{Name: "⏰ Time", Value: ts, Inline: true},
			{Name: "🆔 Session", Value: sessionDisplay, Inline: true},
			locationField,
		}
	default:
		title = "Claude Code Event"
		description = "Claude Code event triggered"
		fields = []discordField{
			{Name: "⏰ Time", Value: ts, Inline: true},
			{Name: "📋 Event", Value: hookType, Inline: true},
			{Name: "🆔 Session", Value: sessionDisplay, Inline: true},
			locationField,
		}
	}

	return discordEmbed{
		Title:       title,
		Description: description,
		Color:       color,
		Timestamp:   isoTS,
		Footer:      map[string]string{"text": "Project • " + projectName},
		Fields:      fields,
	}
}

File: hooks/scripts/_archived_ts/notifications/providers/discord.ts
// Discord notification provider — sends rich embed messages to Discord webhooks
import * as path from "node:path";
import { send } from "../lib/sender.ts";
import type { SendResult } from "../lib/sender.ts";
import type { HookInput } from "./telegram.ts";

const COLORS = {
  Stop: 5763719,          // Green
  SubagentStop: 3447003,  // Blue
  AskUserPrompt: 15844367, // Yellow
  default: 10070709,      // Gray
} as const;

function getProjectName(cwd?: string): string {
  if (!cwd) return "Unknown";
  return path.basename(cwd) || "Unknown";
}

function formatTimestamp(): string {
  return new Date().toLocaleTimeString("en-US", {
    hour: "2-digit", minute: "2-digit", second: "2-digit", hour12: false,
  });
}

function truncateSessionId(sessionId?: string): string {
  if (!sessionId) return "N/A";
  return sessionId.length > 8 ? `${sessionId.slice(0, 8)}...` : sessionId;
}

interface DiscordEmbed {
  title: string;
  description: string;
  color: number;
  timestamp: string;
  footer: { text: string };
  fields: { name: string; value: string; inline: boolean }[];
}

function buildEmbed(input: HookInput): DiscordEmbed {
  const hookType = input.hook_event_name || "unknown";
  const cwd = input.cwd || "";
  const sessionId = input.session_id || "";
  const projectName = getProjectName(cwd);
  const color = (COLORS as Record<string, number>)[hookType] ?? COLORS.default;
  const sessionDisplay = `\`${truncateSessionId(sessionId)}\``;
  const locationField = { name: "📍 Location", value: `\`${cwd || "Unknown"}\``, inline: false };

  switch (hookType) {
    case "Stop":
      return {
        title: "Claude Code Session Complete",
        description: "Session completed successfully",
        color, timestamp: new Date().toISOString(),
        footer: { text: `Project • ${projectName}` },
        fields: [
          { name: "⏰ Time", value: formatTimestamp(), inline: true },
          { name: "🆔 Session", value: sessionDisplay, inline: true },
          locationField,
        ],
      };

    case "SubagentStop":
      return {
        title: "Claude Code Subagent Complete",
        description: "Specialized agent completed its task",
        color, timestamp: new Date().toISOString(),
        footer: { text: `Project • ${projectName}` },
        fields: [
          { name: "⏰ Time", value: formatTimestamp(), inline: true },
          { name: "🔧 Agent Type", value: input.agent_type || "unknown", inline: true },
          { name: "🆔 Session", value: sessionDisplay, inline: true },
          locationField,
        ],
      };

    case "AskUserPrompt":
      return {
        title: "Claude Code Needs Input",
        description: "Claude is waiting for user input",
        color, timestamp: new Date().toISOString(),
        footer: { text: `Project • ${projectName}` },
        fields: [
          { name: "⏰ Time", value: formatTimestamp(), inline: true },
          { name: "🆔 Session", value: sessionDisplay, inline: true },
          locationField,
        ],
      };

    default:
      return {
        title: "Claude Code Event",
        description: "Claude Code event triggered",
        color, timestamp: new Date().toISOString(),
        footer: { text: `Project • ${projectName}` },
        fields: [
          { name: "⏰ Time", value: formatTimestamp(), inline: true },
          { name: "📋 Event", value: hookType, inline: true },
          { name: "🆔 Session", value: sessionDisplay, inline: true },
          locationField,
        ],
      };
  }
}

export const discord = {
  name: "discord",
  isEnabled: (env: Record<string, string>) => !!env.DISCORD_WEBHOOK_URL,
  send: async (input: HookInput, env: Record<string, string>): Promise<SendResult> => {
    if (!env.DISCORD_WEBHOOK_URL) return { success: false, error: "DISCORD_WEBHOOK_URL not configured" };
    return send("discord", env.DISCORD_WEBHOOK_URL, { embeds: [buildEmbed(input)] });
  },
};
